### PR TITLE
Output linear matrices

### PR DESCRIPTION
**Pull request by markdewing**
## Proposed changes

Output linear method matrices (overlap and Hamiltonian) in scalar.dat format.
They can be processed with qmca and check_scalars.py, and compared between the batched and non-batched versions of the code.
A separate option (`freeze_parameters`) disables updating the parameters so the output can be used for statistics.

Enable by adding the parameters `output_matrices` and `freeze_parameters` to the qmc block
```
   <qmc method="linear_batch" move="pbyp" checkpoint="-1">
      <parameter name="output_matrices">yes</parameter>
      <parameter name="freeze_parameters">yes</parameter>

```

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
laptop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by markdewing** [comment]
Attached is a script to compare the output of qmca for two runs.

[compare_qmca.zip](https://github.com/QMCPACK/qmcpack/files/5860774/compare_qmca.zip)

**Comment by prckent** [comment]
Very useful to have.

Just one concern: I think that the freezing/non-updating of parameters also should be well advertised. Perhaps output_matrices should be output_matrices_freeze_parameters or output_matrices_no_update? debug_matrices? A warning around line 1294 of QMCFixedSampleLinearOptimize would also help.

**Comment by markdewing** [comment]
If the effect of freezing the parameters is a concern, it could be moved to a separate attribute - "freeze_parameters".

**Comment by jtkrogel** [comment]
Why an xml attribute vs a `<parameter>`?  The qmc elements have all reached such a size that line-listed parameters are the common, and IMO appropriate, case.

**Comment by markdewing** [comment]
Using an attribute was lazily extending the existing code. A `<parameter>` is probably better.

**Comment by prckent** [comment]
@markdewing Separate parameters is a good.

**Review by prckent — APPROVED** [review]

File: src/QMCDrivers/WFOpt/OutputMatrix.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+//
+// File created by: Mark Dewing, mdewing@anl.gov, Argonne National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+
+/** @file OutputMatrix.cpp
+ * @brief Output the Hamiltonian and overlap matrices from linear method
+ */
+
+#include "QMCDrivers/WFOpt/OutputMatrix.h"
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/OutputMatrix.cpp:17]: OutputMatrix.cpp needs to include "OutputMatrix.h". Related header files should be included without any path.

**Review by ye-luo — COMMENTED** [review]
> A separate option (freeze_parameters) disables updating the parameters so the output can be used for statistics.

What statistics? I'm confused.

**Comment by markdewing** [comment]
The output of the matrix files is in scalar.dat format that can be post-processed with qmca or check_scalars.py.    Each iteration of the "loop" construct in the input adds another line to the file - the freeze_parameters option keeps the parameters from changing at each iteration.

**Comment by ye-luo** [comment]
If it is frozen, why running multiple loop iterations is needed? What is the intention?

**Comment by markdewing** [comment]
The intention is to compare between the old,non-batched code and the batched code, and for regression tests.  It started with parameter gradients.  The existing 'test' input (in Optimize/testDerivOptimization.cpp) would print out the analytic gradient and numerical gradient.   I extended it to also output the analytic gradients to a scalar.dat file.   This can be used for regression testing, for instance, in tests/molecules/He_param.   In order for check_scalars.py to generate statistics, there needs to be multiple lines in the *.scalar.dat input file (of parameter gradients).    The gradient evaluation is carried out once per optimizer loop iteration.  In order to get multiple lines in the parameter gradient file, I hacked it to print one line per optimizer loop.

The input file to the He_param tests has the lines
```
  <loop max="10">
    <qmc method="opt_batch" move="pbyp" checkpoint="-1" gpu="no">
      <optimize method="test" output_param_file="yes"/>
```
This produces 10 lines of parameter gradients in the output file (He.param.s000.scalar.dat).
It's a hack.  Maybe there's a better way of doing this?

I extended the same pattern to the H and S matrices for the linear method.

**Comment by ye-luo** [comment]
Now I get the point, even if parameters are fronzen, samples are changing and so the statistics are changing.
1. loop should update s00x. So you can get one file per iteration.
2. Checking the average of 10 loop iteration is more for convenience, it doesn't correspond to any real needs.
3. run_qmc can check scalar.dat at every QMC section. So loop=2 and check two scalar.dat even with just one line each should be sufficient to cover the real use case, isn't it?

**Pull request merged**